File: overlay/tests/test_dyn_maxbytes.c
/*
 * test_dyn_maxbytes.c - Test dynamically resizing maximum memory limit.
 */

#include "c_harness.h"

static int test_resize_maxbytes(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char line[256];
  test_uint64_t maxbytes = 0;

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_send_str(&c, "cache_memlimit 128\r\n") == 0);
  TEST_ASSERT(client_readline(&c, line, sizeof(line)) > 0);
  TEST_ASSERT(strncmp(line, "OK", 2) == 0);

  TEST_ASSERT(client_send_str(&c, "stats settings\r\n") == 0);
  while (client_readline(&c, line, sizeof(line)) > 0) {
    if (strncmp(line, "END", 3) == 0) {
      break;
    }
    if (strstr(line, "STAT maxbytes ") != NULL) {
      unsigned long long val = 0;
      if (sscanf(line, "STAT maxbytes %llu", &val) == 1) {
        maxbytes = (test_uint64_t)val;
      }
    }
  }

  TEST_ASSERT(maxbytes == (test_uint64_t)128 * 1024 * 1024);

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

int test_dyn_maxbytes_run(void) {
  RUN_TEST(test_resize_maxbytes);
  return 0;
}

File: overlay/tests/test_dyn_maxbytes.py
"""
test_dyn_maxbytes.py - Test dynamically resizing maximum memory limit.
"""

import os
import sys
import unittest

sys.path.insert(0, os.path.dirname(__file__))
from memcached_harness import MemcachedServer


class TestDynMaxbytes(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.server = MemcachedServer().start()

    @classmethod
    def tearDownClass(cls):
        cls.server.stop()

    def setUp(self):
        self.mc = self.server.client()

    def tearDown(self):
        self.mc.close()

    def test_resize_maxbytes(self):
        st = self.mc.stats("settings")
        orig_maxbytes = int(st["maxbytes"])

        # Change maxbytes to 128MB using cache_memlimit command
        self.mc.send_cmd("cache_memlimit 128")
        res = self.mc.readline().strip()
        self.assertEqual(res, "OK")

        st2 = self.mc.stats("settings")
        self.assertEqual(int(st2["maxbytes"]), 128 * 1024 * 1024)


if __name__ == "__main__":
    unittest.main()
